etl/src/replication/apply.rs

```diff
@@ -429,143 +504,199 @@ where
+
+            // If we have an event, and we want to keep it, we add it to the batch and update the last
+            // commit lsn (if any).
+            let should_include_event = matches!(result.end_batch, None | Some(EndBatch::Inclusive));
+            if let Some(event) = result.event
+                && should_include_event
+            {
+                state.events_batch.push(event);
+                state.update_last_commit_end_lsn(result.end_lsn);
+            }
 
-    if let Some(event) = result.event
-        && matches!(result.end_batch, None | Some(EndBatch::Inclusive))
-    {
-        state.events_batch.push(event);
-        state.update_last_commit_end_lsn(result.end_lsn);
-    }
+            let mut continue_loop = true;
+
+            // If we have elements in the batch, and we have reached the max batch size, or we are told
+            // to end the batch, we send it.
+            if state.events_batch.len() >= max_batch_size || result.end_batch.is_some() {
+                // We check if the batch has elements. It can be that a batch has no elements when
+                // the batch is ended prematurely, and it contains only skipped events. In this case,
+                // we don't produce any events to the destination but downstream code treats it as if
+                // those evets are "persisted".
+                if !state.events_batch.is_empty() {
+                    send_batch(state, events_stream.as_mut(), destination, max_batch_size).await?;
+                }
 
-    try_send_batch(
-        state,
-        result.end_batch,
-        result.table_replication_error,
-        destination,
-        hook,
-        max_batch_size,
-        max_batch_fill_duration,
-    )
-    .await
+                // If we have a caught table error, we want to mark the table as errored.
+                //
+                // Note that if we have a failure after marking a table as errored and events will
+                // be reprocessed, even the events before the failure will be skipped.
+                //
+                // Usually in the apply loop, errors are propagated upstream and handled based on if
+                // we are in a table sync worker or apply worker, however we have an edge case (for
+                // relation messages that change the schema) where we want to mark a table as errored
+                // manually, not propagating the error outside the loop, which is going to be handled
+                // differently based on the worker:
+                // - Apply worker -> will continue the loop skipping the table.
+                // - Table sync worker -> will stop the work (as if it had a normal uncaught error).
+                // Ideally we would get rid of this since it's an anomalous case which adds unnecessary
+                // complexity.
+                if let Some(error) = result.table_replication_error {
+                    continue_loop &= hook.mark_table_errored(error).await?;
```

Comment:
Decided to use `continue_loop` instead of `end_loop` to avoid confusing negations.